storage/v1alpha1/types_share.go

```diff
@@ -0,0 +1,149 @@
+// ClusterRole and ClusterRoleBinding.
+//
+// apiVersion: rbac.authorization.k8s.io/v1
+// kind: ClusterRole
+// metadata:
+//   name: shared-resource-my-share
+// rules:
+//   - apiGroups:
+//       - storage.openshift.io
+//     resources:
+//       - sharedresources
+//     resourceNames:
+//       - my-share
+//     verbs:
+//       - get
+//       - list
+//       - watch
+//
+// apiVersion: rbac.authorization.k8s.io/v1
+// kind: ClusterRoleBinding
+// metadata:
+//   name: shared-resource-my-share
+// roleRef:
+//   apiGroup: rbac.authorization.k8s.io
+//   kind: ClusterRole
+//   name: shared-resource-my-share
+// subjects:
+//   - kind: ServiceAccount
+//     name: default
+//     namespace: my-csi-app-namespace
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// These capabilities should not be used by applications needing long term support.
+// +k8s:openapi-gen=true
+// +openshift:compatibility-gen:level=4
+// +kubebuilder:subresource:status
+//
+type SharedResource struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Specification of the desired shared resource
+	// +kubebuilder:validation:Required
+	Spec SharedResourceSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+
+	// Observed status of the shared resource
+	Status SharedResourceStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// SharedResourceList contains a list of SharedResource objects.
+//
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+type SharedResourceList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	Items           []SharedResource `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// SharedResourceSpec defines the desired state of a SharedResource.
+// +k8s:openapi-gen=true
+type SharedResourceSpec struct {
+	// resource references the backing object for this shared resource.
+	// +kubebuilder:validation:Required
+	Resource ResourceReference `json:"resource" protobuf:"bytes,1,name=resource"`
+
+	// description is a user readable explanation of what the backing resource provides.
+	Description string `json:"description,omitempty" protobuf:"bytes,2,opt,name=description"`
+}
+
+// SharedResourceStatus contains the observed status of shared resource. Read-only.
+type SharedResourceStatus struct {
```

Comment:
Is there an efficient way for a cluster-admin to know whether a particular sharedresource is in use?   You cannot keep a count here because it would be too expensive to update, but how can cluster-admin know that it is safe to delete a sharedresource?

Reply:
we plan to add metrics that will include shares in use, how many pods are using a given share

do we need that level of detail in the godoc, or are you just asking out of long term interest?

Reply:
This perhaps another item to address as we go from tech preview to GA.

Reply:
> This perhaps another item to address as we go from tech preview to GA.

This is a requirement for observability.  Without an ability to determine usage, how will you make an informed decision about whether you can promote your tech preview API to GA?